File: packages/py/application/src/bolsa_application/order_proposal_mapper.py
"""Proyección decision_session (kind=propose) → OrderProposalV1 (refs-only)."""

from __future__ import annotations

from typing import Any, Literal

from bolsa_domain.entities.cognitive_artifacts import DecisionSessionRecord

OrderProposalStatus = Literal["open", "confirmed", "rejected", "superseded", "expired"]

_REC_TO_PROPOSAL: dict[str, OrderProposalStatus] = {
    "approved": "confirmed",
    "rejected": "rejected",
    "expired": "expired",
    "superseded": "superseded",
}


def _derive_proposal_status(session: DecisionSessionRecord) -> OrderProposalStatus:
    payload = session.payload or {}
    rec = payload.get("recommendation")
    rec_status = ""
    if isinstance(rec, dict):
        rec_status = str(rec.get("status") or "")
    mapped = _REC_TO_PROPOSAL.get(rec_status)
    if mapped is not None:
        return mapped
    if session.status == "closed":
        return "confirmed"
    return "open"


def session_to_order_proposal(session: DecisionSessionRecord) -> dict[str, Any] | None:
    """Maps a propose session row to OrderProposalV1 dict; None if not mappable."""
    if session.kind != "propose":
        return None

    payload = session.payload or {}
    decision_id = session.decision_id or payload.get("decisionId")
    recommendation_id = session.recommendation_id
    if recommendation_id is None and isinstance(payload.get("recommendation"), dict):
        recommendation_id = payload["recommendation"].get("recommendationId")

    if not decision_id or not recommendation_id:
        return None

    status = _derive_proposal_status(session)
    closed_at: str | None = None
    if status != "open" and session.status == "closed":
        closed_at = session.created_at

    return {
        "artifactType": "ART-ORDER-PROPOSAL",
        "schemaVersion": "1.0.0",
        "proposalId": session.id,
        "decisionId": str(decision_id),
        "recommendationId": str(recommendation_id),
        "sessionId": session.id,
        "accountId": session.account_id,
        "instrumentId": session.instrument_id,
        "status": status,
        "createdAt": session.created_at,
        "closedAt": closed_at,
    }

File: packages/shared/src/cognitive/order-proposal-mapper.ts
/**
 * Proyecta DecisionSessionV1 kind=propose → OrderProposalV1 (refs-only).
 */

import type { DecisionSessionV1 } from "./decision-session.js";
import type { OrderProposalStatus, OrderProposalV1 } from "./order-proposal.js";

const PROPOSAL_STATUSES: ReadonlySet<string> = new Set([
  "open",
  "confirmed",
  "rejected",
  "superseded",
  "expired",
]);

function deriveProposalStatus(session: DecisionSessionV1): OrderProposalStatus {
  const recStatus =
    session.recommendation &&
    typeof session.recommendation === "object" &&
    "status" in session.recommendation
      ? String((session.recommendation as { status?: string }).status ?? "")
      : "";

  if (recStatus === "approved") return "confirmed";
  if (recStatus === "rejected") return "rejected";
  if (recStatus === "expired") return "expired";
  if (recStatus === "superseded") return "superseded";
  if (session.status === "closed") return "confirmed";
  return "open";
}

/** Maps a propose session to OrderProposalV1; returns null if not mappable. */
export function sessionToOrderProposal(
  session: DecisionSessionV1,
): OrderProposalV1 | null {
  if (session.kind !== "propose") return null;

  const decisionId = session.decisionId ?? "";
  const recommendationId =
    session.recommendationId ??
    (session.recommendation &&
    typeof session.recommendation === "object" &&
    "recommendationId" in session.recommendation
      ? String(
          (session.recommendation as { recommendationId?: string })
            .recommendationId ?? "",
        )
      : "");

  if (!decisionId || !recommendationId) return null;

  const status = deriveProposalStatus(session);
  const closedAt =
    status !== "open" && session.status === "closed" ? session.createdAt : null;

  return {
    artifactType: "ART-ORDER-PROPOSAL",
    schemaVersion: "1.0.0",
    proposalId: session.sessionId,
    decisionId,
    recommendationId,
    sessionId: session.sessionId,
    accountId: session.accountId ?? null,
    instrumentId: session.instrumentId,
    status: PROPOSAL_STATUSES.has(status) ? status : "open",
    createdAt: session.createdAt,
    closedAt,
  };
}
